### ISSUE TITLE
Performance issues related to undo/redo and history

### ISSUE THREAD

**Issue by tigrouind** [LDmicro/LDmicro#245]
**Describe the bug**
After inserting/deleting a few contacts on a LD file of reasonable size (eg: 50 rungs, 100 contacts), LDmicro becomes slow.
As a result, any operation (e.g. inserting, editing, deleting a contact) takes at least one second, whereas under normal circumstances (e.g. after a new program start) this operation is instantaneous.

**To Reproduce**
1. Start LDmicro
2. Open file inside [test.zip](https://github.com/LDmicro/LDmicro/files/10950984/test.zip)
3. Repeat the following operations until it slows down :
- press C (once) to insert a contact
- press Delete (once) to delete a contact

**Expected behaviour**
LDmicro remains responsive and fast (most operations should be instantaneous).

**Desktop**
 - OS: Windows 10 22H2
 - Version 5.4.1.1

---

EDIT : I ran LDmicro under a debugger. When it freezes, most CPU time is spent in [ForgetFromGrid()](https://github.com/LDmicro/LDmicro/blob/master/ldmicro/schematic.cpp#L309)
This function is called for every element to be freed. For some reason, some elements to be removed are not in `DisplayMatrix` array, and it will loop  256 * 20000 = 5120000 times (for each element). I think elements are not found because they have already been removed before (they only exist in the undo history) and got replaced by a newer version (what is currently show on the screen at that grid location).

Here is full stack trace :

```
CHANGING_PROGRAM(x)
UndoRemember()
Undo::pushUndo()
ProgramStack::push()
UndoStruct::reset() //only call that method if there is 32 items in history
PlcProgram::~PlcProgram()
PlcProgram::reset() 
FreeCircuit()
ForgetFromGrid() //slow code (only if element to be removed is not found)
```

The issue only occurs after several insertion/deletion because undo history must reach [32 items](https://github.com/LDmicro/LDmicro/blob/master/ldmicro/undoredo.cpp#L124)

Is ForgetFromGrid() still necessary, considering :
- DisplayMatrix array is [cleared](https://github.com/LDmicro/LDmicro/blob/master/ldmicro/draw_outputdev.cpp#L279) every time circuit is draw on the screen.
- The [function](https://github.com/LDmicro/LDmicro/blob/master/ldmicro/circuit.cpp#L1367) that is supposed to clear out that array (eg: after circuit has been modified) has been commented out.

**Comment by parksplit** [comment]
is there a way to fix this issue?

**Comment by tigrouind** [comment]
If you looking for a quick fix / workaround, you could lower DISPLAY_MATRIX_X_SIZE and DISPLAY_MATRIX_Y_SIZE back to their [original values](https://github.com/LDmicro/LDmicro/blob/8fc0cc7efaf917641ff2cdcb114a7fd7847215a2/ldmicro/ldmicro.h#L589).

The size of the circuit will be limited to 16 columns and 512 rows but that should be enough for most projects (can be slightly increased if needed).

For a proper fix, investigation need to be made. First to check if bug has been introduced at some point or existed since beginning. The main issue is that ForgetFromGrid is called while the element to be freed is not more in DisplayMatrix array. Was it already like in initial commit ?

**Comment by parksplit** [comment]
Hi tigrouind,
I hope this message finds you well. I wanted to ask if it would be possible to build LDmicro version 5.4.1.1 with the original values for DISPLAY_MATRIX_X_SIZE and DISPLAY_MATRIX_Y_SIZE that you proposed.
I believe reverting to those original settings might help address the performance issues we've been discussing.
Thank you for your efforts on this project, and I look forward to your response.
Best regards,
Parksplit

**Comment by tigrouind** [comment]
@parksplit Here is a [32-bit Windows build](https://github.com/user-attachments/files/16575748/ldmicro.zip)

Decreasing DISPLAY_MATRIX_X_SIZE and DISPLAY_MATRIX_Y_SIZE solves the issue (as expected). It put 32 x 512. It seems that DISPLAY_MATRIX_X_SIZE is the maximum number of elements per row but empty slots in a row also count as elements.
When the window is maximized, 16 elements per row cause an error on my 2560x1440 display.

**Comment by parksplit** [comment]
Thank you very much for this 32 bit version, I´m going to test it.

Also: Can I share this version to the ldmicro community forum? I think there might be people with the same problem.

**Comment by tigrouind** [comment]
Yes, you can share it on the forum. Btw, I have created another version (see my previous zip file) that do not call ForgetFromGrid() function anymore (with 256 x 9999 limits). As stated in my first message, I don't think that method is needed anymore. I made some tests and it seems to work.

**Comment by tigrouind** [comment]
If anybody want to build this from the sources, here is the respective fixes : 
- [undo_redo_fix1](https://github.com/tigrouind/LDmicro/commit/a4661d007a87fdff9519d6f1c5c1f8e120158abf)
- [undo_redo_fix2](https://github.com/tigrouind/LDmicro/commit/73b910b3abf2f9120cfdb1c62ecd157106785681)
